src/rgw/rgw_client_io.h

```diff
@@ -4,83 +4,236 @@
+   * code base between dynamic and static decorators. The difference is about
+   * what we store internally: pointer to a decorated object versus the whole
+   * object itself. */
+  template <typename T = void,
+            typename std::enable_if<
+    std::is_pointer<DecorateeT>::value, T>::type* = nullptr>
+  DerefedDecorateeT& get_decoratee() {
+    return *decoratee;
+  }
+
+  template <typename T = void,
+            typename std::enable_if<
+    ! std::is_pointer<DecorateeT>::value, T>::type* = nullptr>
+  DerefedDecorateeT& get_decoratee() {
+    return decoratee;
+  }
 
 protected:
-  virtual int write_data(const char *buf, int len) = 0;
-  virtual int read_data(char *buf, int max) = 0;
+  void init_env(CephContext *cct) override {
+    return get_decoratee().init_env(cct);
+  }
 
 public:
-  virtual ~RGWStreamIO() {}
-  RGWStreamIO() : bytes_sent(0), bytes_received(0), sha256_hash(nullptr) {}
+  RGWDecoratedRestfulIO(DecorateeT&& decoratee)
+    : decoratee(std::move(decoratee)) {
+  }
 
-  int print(const char *format, ...);
-  int write(const char *buf, int len);
-  virtual void flush() = 0;
-  int read(char *buf, int max, int *actual, bool hash = false);
+  std::size_t send_status(const int status,
+                          const char* const status_name) override {
+    return get_decoratee().send_status(status, status_name);
+  }
+
+  std::size_t send_100_continue() override {
+    return get_decoratee().send_100_continue();
+  }
+
+  std::size_t send_header(const boost::string_ref& name,
+                          const boost::string_ref& value) override {
+    return get_decoratee().send_header(name, value);
+  }
+
+  std::size_t send_content_length(const uint64_t len) override {
+    return get_decoratee().send_content_length(len);
+  }
+
+  std::size_t send_chunked_transfer_encoding() override {
+    return get_decoratee().send_chunked_transfer_encoding();
+  }
+
+  std::size_t complete_header() override {
+    return get_decoratee().complete_header();
+  }
+
+  std::size_t recv_body(char* const buf, const std::size_t max) override {
+    return get_decoratee().recv_body(buf, max);
+  }
+
+  std::size_t send_body(const char* const buf,
+                        const std::size_t len) override {
+    return get_decoratee().send_body(buf, len);
+  }
+
+  void flush() override {
+    return get_decoratee().flush();
+  }
+
+  RGWEnv& get_env() noexcept override {
+    return get_decoratee().get_env();
+  }
+
+  int complete_request() override {
+    return get_decoratee().complete_request();
+  }
+};
+
+
+/* We're doing this nasty thing only because of extensive usage of templates
+ * to implement the static decorator pattern. C++ templates de facto enforce
+ * mixing interfaces with implementation. Additionally, those classes derive
+ * from RGWRestfulIO defined here. I believe that including in the middle of
+ * file is still better than polluting it directly. */
+#include "rgw_client_io_decoimpl.h"
```

Comment:
this in unfortunate, but i definitely think it's worth the tradeoff. and thanks for the explanation